zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,742 @@
 		(struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+	int i;
 
-	for (size_t i = 0;
-		i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
-		i++) {
-		vm_heaps[i].virtual_region = &virtual_memory_regions[i];
+	/* Check if we haven't created heap for this region already */
+	if (vmh_get_heap_by_attribute(memory_region_attribute, core_id))
+		return NULL;
 
-		switch (virtual_memory_regions[i].attr)	{
-		case MEM_REG_ATTR_CORE_HEAP:
-			vm_heaps[i].memory_caps =
-				SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP | SOF_MEM_CAPS_CACHE;
-			break;
-		case MEM_REG_ATTR_SHARED_HEAP:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+	struct virtual_memory_heap *new_heap =
+		rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*new_heap));
+
+	if (!new_heap)
+		return NULL;
+
+	list_init(&new_heap->node);
+	struct vmh_heap_config new_config = {0};
+
+	/* Search for matching attribute so we place heap on right
+	 * virtual region.
+	 * In case of core heaps regions we count theme from 0 to max
+	 * available cores
+	 */
+	if (memory_region_attribute == MEM_REG_ATTR_CORE_HEAP) {
+		new_heap->virtual_region = &virtual_memory_regions[core_id];
+	} else {
+		for (i = CONFIG_MP_MAX_NUM_CPUS;
+			i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT; i++) {
+
+			if (memory_region_attribute == virtual_memory_regions[i].attr) {
+				new_heap->virtual_region = &virtual_memory_regions[i];
+				break;
+			}
+		}
+	}
+
+	if (!new_heap->virtual_region)
+		goto fail;
+
+	/* If no config was specified we will use default one */
+	if (!cfg) {
+		vmh_get_default_heap_config(new_heap->virtual_region, &new_config);
+		cfg = &new_config;
+	}
+
+	/* Check if configuration provided can fit into virtual regions memory
+	 * and if it is physical page aligned
+	 */
+	size_t total_requested_size = 0;
+
+	for (i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		if (cfg->block_bundles_table[i].block_size > 0 &&
+			is_power_of_2(cfg->block_bundles_table[i].block_size)) {
+			total_requested_size += cfg->block_bundles_table[i].block_size *
+						cfg->block_bundles_table[i].number_of_blocks;
+
+			if (total_requested_size > new_heap->virtual_region->size ||
+					total_requested_size % CONFIG_MM_DRV_PAGE_SIZE)
+				goto fail;
+		}
+	}
+
+	/* Offset used to calculate where in virtual region to start buffer
+	 * for the memblock
+	 */
+	uint32_t offset = 0;
+
+	/* Iterate over all blocks and fill those that have config size value
+	 * higher than 0.
+	 * Each cfg entry is converted into new allocators based on mem_blocks
+	 * API. It needs a lot of runtime allocations for members of mem_blocks.
+	 * The allocation_sizes bit arrays are runtime allocated too.
+	 */
+	for (i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		/* Only run for sizes > 0 */
+		if (!cfg->block_bundles_table[i].block_size)
+			continue;
+
+		size_t bitarray_size = SOF_DIV_ROUND_UP(
+			SOF_DIV_ROUND_UP(cfg->block_bundles_table[i].number_of_blocks, 8),
```

Comment:
why is 8 special here ?

Reply:
It is a part of calculation of sys_bitarray.num_bundles that I applied from
```
/**
 * @brief Create a bitarray object.
 *
 * @param name Name of the bitarray object.
 * @param total_bits Total number of bits in this bitarray object.
 * @param sba_mod Modifier to the bitarray variables.
 */
#define _SYS_BITARRAY_DEFINE(name, total_bits, sba_mod)			\
	sba_mod uint32_t _sys_bitarray_bundles_##name			\
		[DIV_ROUND_UP(DIV_ROUND_UP(total_bits, 8),		\
			       sizeof(uint32_t))] = {0};		\
	sba_mod sys_bitarray_t name = {					\
		.num_bits = total_bits,					\
		.num_bundles = DIV_ROUND_UP(				\
			DIV_ROUND_UP(total_bits, 8), sizeof(uint32_t)),	\
		.bundles = _sys_bitarray_bundles_##name,		\
	}

```
number itself represents number of bits in a byte :P So I did not want to do a define like that.
Whole recipe in this line calculates number of bundles for the bit array.

Reply:
ok, lets do an inline comment here so that others can understand that 8 is for bits in byte (and not any magic)